rln/src/protocol.rs

```diff
@@ -374,13 +370,81 @@ pub enum ProofError {
     SynthesisError(#[from] SynthesisError),
 }
 
+fn calculate_witness_element<E: ark_ec::PairingEngine>(
+    witness: Vec<BigInt>,
+    sanity_check: bool,
+) -> Result<Vec<E::Fr>> {
+    use ark_ff::{FpParameters, PrimeField};
+    let modulus = <<E::Fr as PrimeField>::Params as FpParameters>::MODULUS;
+
+    // convert it to field elements
+    use num_traits::Signed;
+    let witness = witness
+        .into_iter()
+        .map(|w| {
+            let w = if w.sign() == num_bigint::Sign::Minus {
+                // Need to negate the witness element if negative
+                modulus.into() - w.abs().to_biguint().unwrap()
+            } else {
+                w.to_biguint().unwrap()
+            };
+            E::Fr::from(w)
+        })
+        .collect::<Vec<_>>();
+
+    Ok(witness)
+}
```

Comment:
This function was extracted from [ark-circom](https://github.com/gakonst/ark-circom/blob/master/src/witness/witness_calculator.rs#L258-L286), It's main difference is that it will receive the witness instead of calculating it inside the function.